File: modules/tracker/call_api.py
import orjson
import traceback
from typing import Optional
from aiohttp import ClientSession, ClientResponse

from modules.enums import SekaiServerRegion
from ..schema.call_api import Top100RankingResponse, BorderRankingResponse


class HarukiSekaiAPIClient:
    def __init__(self, api_endpoint: str) -> None:
        self.api_endpoint = api_endpoint
        self.session: Optional[ClientSession] = None
        self.headers = {"User-Agent": "Haruki Event Tracker / v1.0.0"}

    async def init(self) -> None:
        self.session = ClientSession()

    async def close(self) -> None:
        await self.session.close()

    async def _request(self, url: str) -> Optional[bytes]:
        async with self.session.get(url, headers=self.headers, timeout=20) as response:
            response.raise_for_status()
            if response.status == 200:
                return await response.read()
            else:
                return None

    async def get_top100(self, event_id: int, server: SekaiServerRegion) -> Optional[Top100RankingResponse]:
        try:
            url = f"{self.api_endpoint}/{server.value}/user/%user_id/event/{event_id}/ranking?rankingViewType=top100"
            response = await self._request(url)
            return Top100RankingResponse(**orjson.loads(response))
        except Exception:
            traceback.print_exc()
            return None

    async def get_border(self, event_id: int, server: SekaiServerRegion) -> Optional[BorderRankingResponse]:
        try:
            url = f"{self.api_endpoint}/{server.value}/event/{event_id}/ranking-border"
            response = await self._request(url)
            return BorderRankingResponse(**orjson.loads(response))
        except Exception:
            traceback.print_exc()
            return None

File: tracker/callapi.go
package tracker

import (
	"context"
	"fmt"
	"haruki-tracker/config"
	"time"

	"haruki-tracker/utils/model"

	"github.com/bytedance/sonic"
	"github.com/go-resty/resty/v2"
)

type HarukiSekaiAPIClient struct {
	apiEndpoint   string
	authorization string
	client        *resty.Client
}

func NewHarukiSekaiAPIClient(apiEndpoint, authorization string) *HarukiSekaiAPIClient {
	client := resty.New().
		SetTimeout(20*time.Second).
		SetHeader("User-Agent", fmt.Sprintf("Haruki-Event-Tracker/%s", config.Version))
	if authorization != "" {
		client.SetHeader("X-Haruki-Sekai-Token", authorization)
	}
	return &HarukiSekaiAPIClient{
		apiEndpoint:   apiEndpoint,
		authorization: authorization,
		client:        client,
	}
}

func (c *HarukiSekaiAPIClient) GetTop100(ctx context.Context, eventID int, server model.SekaiServerRegion) (*model.Top100RankingResponse, error) {
	url := fmt.Sprintf("%s/%s/event/%d/ranking-top100", c.apiEndpoint, server, eventID)
	resp, err := c.client.R().
		SetContext(ctx).
		Get(url)
	if err != nil {
		return nil, fmt.Errorf("failed to get top100: %w", err)
	}
	if resp.StatusCode() != 200 {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode())
	}
	var response model.Top100RankingResponse
	if err := sonic.Unmarshal(resp.Body(), &response); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}
	return &response, nil
}

func (c *HarukiSekaiAPIClient) GetBorder(ctx context.Context, eventID int, server model.SekaiServerRegion) (*model.BorderRankingResponse, error) {
	url := fmt.Sprintf("%s/%s/event/%d/ranking-border", c.apiEndpoint, server, eventID)
	resp, err := c.client.R().
		SetContext(ctx).
		Get(url)
	if err != nil {
		return nil, fmt.Errorf("failed to get border: %w", err)
	}
	if resp.StatusCode() != 200 {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode())
	}
	var response model.BorderRankingResponse
	if err := sonic.Unmarshal(resp.Body(), &response); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}
	return &response, nil
}
